### PR TITLE
[Dubbo-2999] Fixes #2999

### LINKED ISSUE #2999: Enhance the java doc of dubbo-registry module.
Enhance the java doc of dubbo-registry module.

Related project: https://github.com/apache/incubator-dubbo/projects/2

### PR DESCRIPTION
**Pull request by kezhenxu94**
## What is the purpose of the change

fixes #2999

## Brief changelog

XXXXX

## Verifying this change

XXXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/incubator-dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/incubator-dubbo/tree/master/dubbo-test).
- [ ] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/incubator-dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by kezhenxu94** [comment]
Reviewers please pay special attention to the method `org.apache.dubbo.registry.redis.RedisRegistry.Notifier#notifyListenersOnJedisInstance`, which is extracted from the original `run` method, 
```java
                if (first) {
                    first = false;
                    Set<String> keys = jedis.keys(service);
                    if (keys != null && !keys.isEmpty()) {
                        for (String s : keys) {
                            doNotify(jedis, s);
                        }
                    }
                    resetSkip();
                }
```

the `first` variable is never changed before this patch, if I missed anything, please let me know.

**Comment by CrazyHZM** [comment]
Please look at this issue.#1917
Your approach is as wrong as the idea at the beginning. Let me discuss this issue first and look at my thoughts. If my idea is correct, I will reopen my pr.

File: dubbo-registry/dubbo-registry-redis/src/main/java/org/apache/dubbo/registry/redis/RedisRegistry.java [hunk 1]
```diff
@@ -199,46 +164,43 @@ private void deferExpired() {
...
-                    }
-                } finally {
-                    jedis.close();
+            try (Jedis jedis = jedisPool.getResource()) {
+                if (jedis.isConnected()) {
+                    return true; // At least one single machine is available.
                 }
-            } catch (Throwable t) {
+            } catch (Throwable ignored) {
             }
```
Comment (CrazyHZM) [inline dubbo-registry/dubbo-registry-redis/src/main/java/org/apache/dubbo/registry/redis/RedisRegistry.java]: Please close `jedis`  Calling the Close method will return the connection object to the object pool

File: dubbo-registry/dubbo-registry-redis/src/main/java/org/apache/dubbo/registry/redis/RedisRegistry.java [hunk 2]
```diff
@@ -500,13 +443,8 @@ public void onMessage(String key, String msg) {
...
-                try {
-                    Jedis jedis = jedisPool.getResource();
-                    try {
-                        doNotify(jedis, key);
-                    } finally {
-                        jedis.close();
-                    }
+                try (Jedis jedis = jedisPool.getResource()) {
+                    doNotify(jedis, key);
                 } catch (Throwable t) { // TODO Notification failure does not restore mechanism guarantee
```
Comment (CrazyHZM) [inline dubbo-registry/dubbo-registry-redis/src/main/java/org/apache/dubbo/registry/redis/RedisRegistry.java]: Same as above

File: dubbo-registry/dubbo-registry-redis/src/main/java/org/apache/dubbo/registry/redis/RedisRegistry.java (hunk 1 above)
Comment (lixiaojiee) [reply dubbo-registry/dubbo-registry-redis/src/main/java/org/apache/dubbo/registry/redis/RedisRegistry.java]: this type of try block will call the close method automatically

File: dubbo-registry/dubbo-registry-redis/src/main/java/org/apache/dubbo/registry/redis/RedisRegistry.java [hunk 3]
```diff
@@ -86,42 +86,15 @@ public RedisRegistry(URL url) {
...
-        if (url.getParameter("time.between.eviction.runs.millis", 0) > 0) {
-            config.setTimeBetweenEvictionRunsMillis(url.getParameter("time.between.eviction.runs.millis", 0));
-        }
-        if (url.getParameter("min.evictable.idle.time.millis", 0) > 0) {
-            config.setMinEvictableIdleTimeMillis(url.getParameter("min.evictable.idle.time.millis", 0));
-        }
+        GenericObjectPoolConfig config = new URLJedisConfigAdapter(url);
 
         String cluster = url.getParameter("cluster", "failover");
         if (!"failover".equals(cluster) && !"replicate".equals(cluster)) {
```
Comment (lixiaojiee) [inline dubbo-registry/dubbo-registry-redis/src/main/java/org/apache/dubbo/registry/redis/RedisRegistry.java]: Would you please replace the 'failover' with Constants.DEFAULT_CLUSTER?

File: dubbo-registry/dubbo-registry-redis/src/main/java/org/apache/dubbo/registry/redis/RedisRegistry.java (hunk 1 above)
Comment (CrazyHZM) [reply dubbo-registry/dubbo-registry-redis/src/main/java/org/apache/dubbo/registry/redis/RedisRegistry.java]: Sorry, I don’t see it clearly, you are right.

File: dubbo-registry/dubbo-registry-redis/src/main/java/org/apache/dubbo/registry/redis/RedisRegistry.java (hunk 3 above)
Comment (CrazyHZM) [reply dubbo-registry/dubbo-registry-redis/src/main/java/org/apache/dubbo/registry/redis/RedisRegistry.java]: I think you can also put the `replicate` into the Constants.java.

Comment (kezhenxu94) [reply dubbo-registry/dubbo-registry-redis/src/main/java/org/apache/dubbo/registry/redis/RedisRegistry.java]: Sure, changes are made as requested

**Comment by kezhenxu94** [comment]
Thanks @carryxyh for reviewing, I've added more comments in the updated commits.

**Pull request closed without merging**